src/cpp/session/modules/SessionPackages.R

```diff
@@ -1226,79 +772,692 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+   if (!identical(.Platform$OS.type, "windows"))
+      libPaths <- .rs.normalizePath(libPaths)
+
+   subset(libPaths, !duplicated(libPaths))
+})
+
+.rs.addFunction( "writeableLibraryPaths", function()
+{
+   uniqueLibraryPaths <- .rs.uniqueLibraryPaths()
+   writeableLibraryPaths <- character()
+   for (libPath in uniqueLibraryPaths)
+      if (.rs.isLibraryWriteable(libPath))
+         writeableLibraryPaths <- append(writeableLibraryPaths, libPath)
+   writeableLibraryPaths
+})
+
+.rs.addFunction("defaultUserLibraryPath", function()
+{
+   unlist(strsplit(Sys.getenv("R_LIBS_USER"), .Platform$path.sep))[1L]
+})
+
+.rs.addFunction("defaultLibraryPath", function()
+{
+  .libPaths()[1]
+})
+
+.rs.addFunction("isPackageHyperlinkSafe", function(packageName)
+{
+   allowed <- setdiff(
+      c(.packages(), "testthat", "rlang", "devtools", "usethis", "pkgload", "pkgdown"), 
+      c("base", "stats", "utils")
+   )
+   .rs.scalar(
+      packageName %in% allowed
+   )
+})
+
+.rs.addFunction("forceUnloadPackage", function(package)
+{
+   tryCatch(
+      withCallingHandlers(
+         .rs.forceUnloadPackageImpl(package),
+         warning = function(w) invokeRestart("muffleWarning")
+      ),
+      error = warning
+   )
+})
+
+.rs.addFunction("forceUnloadPackageImpl", function(package)
+{
+   .rs.beforePackageUnloaded(package)
+   
+   # Figure out where the package was loaded from before unloading it.
+   pkgPath <- if (package %in% loadedNamespaces())
+   {
+      getNamespaceInfo(package, "path")
+   }
+   
+   # Now, try to unload the package. If it's attached, detach it;
+   # if the namespace is loaded, unload it.
+   searchPathName <- paste("package", package, sep = ":")
+   if (searchPathName %in% search())
+   {
+      detach(
+         name = searchPathName,
+         unload = TRUE,
+         character.only = TRUE,
+         force = TRUE
+      )
+   }
+   else if (package %in% loadedNamespaces())
+   {
+      unloadNamespace(package)
+   }
+   
+   # Now, unload a loaded DLL (if any) associated with the package.
+   dllInfo <- getLoadedDLLs()[[package]]
+   if (!is.null(dllInfo) && !is.null(pkgPath))
+   {
+      suppressWarnings(
+         library.dynam.unload(package, pkgPath)
+      )
+   }
+})
+
+.rs.addFunction("packageVersion", function(name, libPath, pkgs)
+{
+   pkgs <- subset(pkgs, Package == name & LibPath == libPath)
+   if (nrow(pkgs) == 1)
+      pkgs$Version
+   else
+      ""
+})
+
+.rs.addFunction( "initDefaultUserLibrary", function()
+{
+  userdir <- .rs.defaultUserLibraryPath()
+  dir.create(userdir, showWarnings = FALSE, recursive = TRUE)
+  .libPaths(c(userdir, .libPaths()))
+})
+
+.rs.addFunction("ensureWriteableUserLibrary", function()
+{
+   if (!.rs.defaultLibPathIsWriteable())
+      .rs.initDefaultUserLibrary()
+})
+
+.rs.addFunction("lastCharacterIs", function(value, ending) {
+   identical(tail(strsplit(value, "")[[1]], n = 1), ending)
+})
+
+.rs.addFunction("findPackage", function(pkgPath)
```

Comment:
New code; basically a small helper to allow us to accept both package names and package paths.